### ISSUE TITLE
feature: shared memory does not work with spawned processes

### ISSUE THREAD

**Issue by Retribution98** [pmodels/mpich#6603]
Hi there,

I would like to run MPICH with spawned processes and using shared memory, but unfortunately it does not work as I expect.
I have prepared a simple Python reproducer with tabular output showing whether shared buffers in different processes are equal or not.
```python
import sys
from mpi4py import MPI

########################
# Threads initializing #
########################

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

parent_comm = MPI.Comm.Get_parent()

if rank == 0 and parent_comm == MPI.COMM_NULL and size == 1:
    nprocs_to_spawn = 3
    args = ["reproducer.py"]
    info = MPI.Info.Create()
    intercomm = MPI.COMM_SELF.Spawn(
        sys.executable,
        args,
        maxprocs=nprocs_to_spawn,
        info=info,
        root=rank,
    )

#######################
# Merge communicators #
#######################

# Just comment the foollowing block if you want to run without communicators merging.
if parent_comm != MPI.COMM_NULL:
    comm = parent_comm.Merge(high=True)
else:
    comm = intercomm.Merge(high=False)
rank = comm.Get_rank()
size = comm.Get_size()

##############################
# Shared memory initializing #
##############################

# I am going to use non-contiguous shared memory, but it's not necessary.
# The problem is the same as in the case of using contiguous shared memory.
info = MPI.Info.Create()
info.Set("alloc_shared_noncontig", "true")
win = MPI.Win.Allocate_shared(40, MPI.BYTE.size, comm=comm, info=info)
buf, itemsize = win.Shared_query(0)

########################
# Change shared memory #
########################

buf[rank*10:rank*10+10] = bytearray([rank+1]*10)
comm.Barrier()

###########################
# Print comaparison table #
###########################

result = comm.allgather(tuple(buf.tobytes()))
column_width = 5

if rank == 0:
    for i in range(size):
        if i == 0:
            print('|' + ' | '.join(['x'.center(column_width)] + [str(j).center(column_width) for j in range(size)]) + '|')
            print('|' + ' | '.join(['-'*column_width for j in range(size + 1)]) + '|')
        print('|' + ' | '.join([str(i).center(column_width)] + [str(result[i]==result[j]).center(column_width) if i != j else '  x  ' for j in range(size)]) + '|')

##############
# Finish MPI #
##############

if not MPI.Is_finalized():
    win.Free()
    MPI.Finalize()
```

Please save it like as `reproducer.py` and run:
```bash
mpiexec -n 1 python reproducer.py
```

I have these results:

1. If I do not merge the Intercommr with the Intracomm, it works fine - the main process has a separate shared memory and the child processes have a combined shared memory whose buffers are equal to each other.

|  x   |   0  |
|----- | -----|
|  0   |   x  |

|  x   |   0   |   1   |   2  |
|----- | ----- | ----- | -----|
|  0   |   x   |  True |  True|
|  1   |  True |   x   |  True|
|  2   |  True |  True |   x  |

2. BUT if I merge the Intercomm with the Intracomm, it does not work at all. All processes have shared memory, but if I change one buffer, the other buffers do not change.

|  x   |   0   |   1   |   2   |   3  |
|----- | ----- | ----- | ----- | -----|
|  0   |   x   | False | False | False|
|  1   | False |   x   | False | False|
|  2   | False | False |   x   | False|
|  3   | False | False | False |   x  

# FYI:

I installed MPICH from conda-forge. You can see the full environment here:

| Name             | Version   | Build              | Channel     |
| ---------------- | --------- | ------------------ | ----------- |
| _libgcc_mutex    | 0.1       |     conda_forge    | conda-forge |
| _openmp_mutex    | 4.5       |           2_gnu    | conda-forge |
| bzip2            | 1.0.8     |      h7f98852_4    | conda-forge |
| ca-certificates  | 2023.7.22 |      hbcca054_0    | conda-forge |
| ld_impl_linux-64 | 2.40      |      h41732ed_0    | conda-forge |
| libffi           | 3.4.2     |      h7f98852_5    | conda-forge |
| libgcc-ng        | 13.1.0    |      he5830b7_0    | conda-forge |
| libgfortran-ng   | 13.1.0    |      h69a702a_0    | conda-forge |
| libgfortran5     | 13.1.0    |      h15d22d2_0    | conda-forge |
| libgomp          | 13.1.0    |      he5830b7_0    | conda-forge |
| libnsl           | 2.0.0     |      h7f98852_0    | conda-forge |
| libsqlite        | 3.42.0    |      h2797004_0    | conda-forge |
| libstdcxx-ng     | 13.1.0    |      hfd8a6a1_0    | conda-forge |
| libuuid          | 2.38.1    |      h0b41bf4_0    | conda-forge |
| libzlib          | 1.2.13    |      hd590300_5    | conda-forge |
| mpi              | 1.0       |           mpich    | conda-forge |
| mpi4py           | 3.1.4     |  py39h32b9844_0    | conda-forge |
| mpich            | 4.1.1     |    h846660c_100    | conda-forge |
| ncurses          | 6.4       |      hcb278e6_0    | conda-forge |
| openssl          | 3.1.1     |      hd590300_1    | conda-forge |
| pip              | 23.2.1    |    pyhd8ed1ab_0    | conda-forge |
| python           | 3.9.16    | h2782a2a_0_cpython | conda-forge |
| python_abi       | 3.9       |          3_cp39    | conda-forge |
| readline         | 8.2       |      h8228510_1    | conda-forge |
| setuptools       | 68.0.0    |    pyhd8ed1ab_0    | conda-forge |
| tk               | 8.6.12    |      h27826a3_0    | conda-forge |
| tzdata           | 2023c     |      h71feb2d_0    | conda-forge |
| wheel            | 0.41.0    |    pyhd8ed1ab_0    | conda-forge |
| xz               | 5.2.6     |      h166bdaf_0    | conda-forge |

**Comment by YarShev** [comment]
@hzhou, any progress here? Is there a timeline for this issue?

**Comment by YarShev** [comment]
@hzhou, just a friendly reminder

**Comment by hzhou** [comment]
Oh sorry. I meant to reply but was distracted and forgot to hit post. :) I'll take a look at this issue this week and give an update on the progress later.

**Comment by YarShev** [comment]
Sounds good, thanks ;)

**Comment by hzhou** [comment]
@YarShev Pull request #6649 fixes this. Let us know if you could test it.

**Comment by YarShev** [comment]
@hzhou, happy to hear that #6649 fixes this issue. It would be great if you could provide us with the steps to build mpich binaries from your branch. Then we could test it both in a single node and in a cluster. Thanks in advance.

**Comment by YarShev** [comment]
@hzhou, would this instruction be enough to get binaries from your branch working - https://unidist.readthedocs.io/en/latest/optimization_notes/mpi.html#mpich?

**Comment by hzhou** [comment]
First checkout from the pull request. The "Code" button on #6649 provides instruction using github cli and github desktop. You may also checkout the code directly on command line via `git clone https://github.com/hzhou/mpich -b 2308_ch4_local`. To build from github checkout -
```
git submodule update --init
./autogen.sh -yaksa-depth=2
./configure --prefix=/PATH/TO/INSTALL
make
make install
```

**Comment by YarShev** [comment]
@hzhou, thanks, will try to test next week.

**Comment by Retribution98** [comment]
Hi @hzhou !

I tried to use your installation instructions, but I caught the following problem:
```bash
./autogen.sh -yaksa-depth=2
Verifying the location of autogen.sh... done
dirname: missing operand
Try 'dirname --help' for more information.
dirname: missing operand
Try 'dirname --help' for more information.
dirname: missing operand
Try 'dirname --help' for more information.
Checking if autotools are in the same location... yes, all in /
Checking for autoconf version... bad autoconf installation
You either do not have autoconf in your path or it is too old (version
2.69 or higher required). You may be able to use

    autoconf --version

Unfortunately, there is no standard format for the version output and
it changes between autotools versions.  In addition, some versions of
autoconf choose among many versions and provide incorrect output).
```

After that I tried to run the next command, but:
```bash
bash: ./configure: No such file or directory
```

Could you help me solve this problem?

**Comment by Retribution98** [comment]
> You need to install autoconf 2.69 or higher on your system in order to run `autogen.sh`. If it is not available from your package manager (`yum install autoconf`, `apt-get install autoconf`), you can download it here: https://www.gnu.org/software/autoconf/
> 
> In addition to autoconf, you may also need to install automake and libtool.

Thank you so much!
I built it and tested these changes successfully. Thanks for this fix!

**Comment by Retribution98** [comment]
Hi everyone!

I tried to test these changes on the cluster and I found the following problem.

```bash
$ mpiexec -n 1 python reproducer.py
Assertion failed in file src/util/mpir_localproc.c at line 181: node_id < num_nodes
/home/ubuntu/source_lib/mpich/lib/libmpi.so.0(+0x3b4f25) [0x7f6466db4f25]
/home/ubuntu/source_lib/mpich/lib/libmpi.so.0(+0x2d7758) [0x7f6466cd7758]
/home/ubuntu/source_lib/mpich/lib/libmpi.so.0(+0x312bc8) [0x7f6466d12bc8]
/home/ubuntu/source_lib/mpich/lib/libmpi.so.0(+0x24330e) [0x7f6466c4330e]
/home/ubuntu/source_lib/mpich/lib/libmpi.so.0(+0x2439f0) [0x7f6466c439f0]
/home/ubuntu/source_lib/mpich/lib/libmpi.so.0(+0x236fd7) [0x7f6466c36fd7]
/home/ubuntu/source_lib/mpich/lib/libmpi.so.0(PMPI_Intercomm_merge+0x264) [0x7f6466af6614]
/home/ubuntu/miniconda3/envs/timedf-mpich/lib/python3.8/site-packages/mpi4py/MPI.cpython-38-x86_64-linux-gnu.so(+0x97ab3) [0x7f6467bb0ab3]
python(PyCFunction_Call+0x52) [0x55da440c6fd2]
python(_PyObject_MakeTpCall+0x3db) [0x55da440b229b]
python(_PyEval_EvalFrameDefault+0x51d7) [0x55da440ae4c7]
python(_PyEval_EvalCodeWithName+0x2f1) [0x55da440a8261]
python(PyEval_EvalCodeEx+0x39) [0x55da4415afd9]
python(PyEval_EvalCode+0x1b) [0x55da4415af9b]
python(+0x1eb929) [0x55da4417b929]
python(+0x1ea923) [0x55da4417a923]
python(+0x9a00f) [0x55da4402a00f]
python(PyRun_SimpleFileExFlags+0x364) [0x55da44029b13]
python(+0x8cfdc) [0x55da4401cfdc]
python(Py_BytesMain+0x39) [0x55da4414e7b9]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f6468229d90]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7f6468229e40]
python(+0x1be6bd) [0x55da4414e6bd]
Abort(1) on node 0: Internal error

```

I also checked the latest version and this exception was not caught so I think this issue needs to be reopened.

I used this reproducer:
```python
import sys
import time
from mpi4py import MPI

########################
# Threads initializing #
########################

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

parent_comm = MPI.Comm.Get_parent()

if rank == 0 and parent_comm == MPI.COMM_NULL and size == 1:
    nprocs_to_spawn = 5
    args = ["reproducer.py"]
    info = MPI.Info.Create()
    info.Set("hosts", "host1,host2")
    intercomm = MPI.COMM_SELF.Spawn(
        sys.executable,
        args,
        maxprocs=nprocs_to_spawn,
        info=info,
        root=rank,
    )
    comm = intercomm.Merge(high=False)

#######################
# Merge communicators #
#######################

# Just comment the foollowing block if you want to run without communicators merging.
if parent_comm != MPI.COMM_NULL:
    comm = parent_comm.Merge(high=True)
rank = comm.Get_rank()
size = comm.Get_size()

host_comm = comm.Split_type(MPI.COMM_TYPE_SHARED)

##############################
# Shared memory initializing #
##############################
buf_len = 10**2

# I am going to use non-contiguous shared memory, but it's not necessary.
# The problem is the same as in the case of using contiguous shared memory.
info = MPI.Info.Create()
info.Set("alloc_shared_noncontig", "true")
win = MPI.Win.Allocate_shared(buf_len * size, MPI.BYTE.size, comm=comm, info=info)
buf, itemsize = win.Shared_query(0)

########################
# Change shared memory #
########################

new_buf = bytearray([rank+1]*buf_len)
t0 = time.perf_counter()
buf[rank*buf_len:(rank+1)*buf_len] = new_buf
t1 = time.perf_counter()
print(t1-t0)
comm.Barrier()

###########################
# Print comaparison table #
###########################

result = comm.allgather(tuple(buf.tobytes()))
column_width = 5

if rank == 0:
    for i in range(size):
        if i == 0:
            print('|' + ' | '.join(['x'.center(column_width)] + [str(j).center(column_width) for j in range(size)]) + '|')
            print('|' + ' | '.join(['-'*column_width for j in range(size + 1)]) + '|')
        print('|' + ' | '.join([str(i).center(column_width)] + [str(result[i]==result[j]).center(column_width) if i != j else '  x  ' for j in range(size)]) + '|')

##############
# Finish MPI #
##############

if not MPI.Is_finalized():
    win.Free()
    MPI.Finalize()
```

When I tried to remove the `Merge` call, it worked as expected. I got this output:
```bash
|  x   |   0  |
|----- | -----|
|  0   |   x  |

|  x   |   0   |   1   |   2   |   3   |   4  |
|----- | ----- | ----- | ----- | ----- | -----|
|  0   |   x   | False |  True | False |  True|
|  1   | False |   x   | False |  True | False|
|  2   |  True | False |   x   | False |  True|
|  3   | False |  True | False |   x   | False|
|  4   |  True | False |  True | False |   x  |
```

**Comment by YarShev** [comment]
@hzhou, @raffenet, please reopen the issue because we don't have the button.

**Comment by hzhou** [comment]
Thanks for the testing. I am reopening the issue. Fix is coming.

**Comment by hzhou** [comment]
@Retribution98 Could you test PR #6663?

**Comment by Retribution98** [comment]
> @Retribution98 Could you test PR #6663?

Thanks for this hot fix.
I tested this PR, the "comm.merge" problem is resolved and my simple reproducer passes successfully.
Unfortunately my more complex script doesn't work with mpich yet, but I need more time to investigate this problem.

**Comment by YarShev** [comment]
@hzhou, thank you for the supporting this. Look forward to seeing a release containing the feature.

**Comment by YarShev** [comment]
@hzhou, could you please tell us which release will contain this feature and when?

**Comment by hzhou** [comment]
It will be available in the 4.2 release. We expect to release 4.2b1 in early November, and GA release at the end of December or early January.
